prime-router/src/main/kotlin/azure/EmailEngineFunction.kt

```diff
@@ -0,0 +1,341 @@
+            .filter{ shouldFire( it ) }
+
+        schedulesToFire.forEach { schedule -> 
+            val orgs: Iterable<String> = getOrgs( schedule );
+
+            logger.info( "processing ${schedule.template}" )
+
+            // get the orgs to fire for
+            orgs.forEach{ org ->
+                    val emails: List<String> = if ( schedule.emails.size > 0) schedule.emails else getEmails(org, logger)
+                    logger.info( "processing ${org}" )
+                    emails.forEach{ email -> 
+                        logger.info( "sending email to ${email}" )
+                        dispatchToSendGrid( schedule.template, listOf(email), logger );
+                    }
+            }
+        }
+    }
+
+    /**
+     * Determine if a schedule should fire
+     * 
+     * @param schedule to check if it should fire
+     * 
+     * @returns true if the given schedule should fire; false otherwise
+     */
+    private fun shouldFire( schedule: EmailSchedule ): Boolean{
+
+        val parser = CronParser( CronDefinitionBuilder.instanceDefinitionFor(CronType.UNIX) );
+        // Get date for last execution
+        val now = ZonedDateTime.now();
+        val executionTime = ExecutionTime.forCron(parser.parse(schedule.cronSchedule));
+        val timeFromLastExecution = executionTime.timeFromLastExecution(now);
+
+        return ( timeFromLastExecution.get().toSeconds() <= 4*60 /* 4 minutes */);
+    }
+
+    /**
+     * Retreives the organization to send the emailschedule to
+     * 
+     * @param schedule the schedule to check organizations against
+     * 
+     * @returns List of organizations for the schedule
+     */
+    private fun getOrgs( schedule: EmailSchedule ): Iterable<String> {
+        return (
+            if (schedule.organizations.size > 0) 
+                schedule.organizations 
+            else 
+                fetchAllOrgs() 
+        )
+    }
+
+    /**
+     * Retrieves the list of all organization supported
+     * 
+     * @returns List of all organizations supported
+     */
+    private fun fetchAllOrgs(): List<String>{    
+        @Suppress( "NEW_INFERENCE_NO_INFORMATION_FOR_PARAMETER" )    
+        return workflowEngine.db.transactReturning {  tx -> 
+            @Suppress( "UNRESOLVED_REFERENCE")
+            val settings = workflowEngine.db.fetchSettings( SettingType.ORGANIZATION, tx )
+            @Suppress( "NEW_INFERENCE_NO_INFORMATION_FOR_PARAMETER" )
+            settings.map{ it.getName() }
+        }
+    }
+
+    /**
+     * Validates the JWT Token supplied in the authorization header
+     * 
+     * @param request the HTTPRequest object
+     * @param logger logger
+     * 
+     * @returns user from the token; otherwise null
+     */
+    fun validateUser( request: HttpRequestMessage<String?>, logger: Logger): String? {
+        var jwtToken = request.headers["authorization"] ?: ""
+
+        jwtToken = if (jwtToken.length > 7) jwtToken.substring(7) else ""
```

Comment:
This does eerily similar things to https://github.com/CDCgov/prime-reportstream/blob/6abe7c7867e3776eafa36fc802fb0cd4a63d4749/prime-router/src/main/kotlin/azure/HistoryFunctions.kt#L521; this is a good candidate to merge these things together to avoid desynchronization between these different places in the future.

I'm fine 'postponing' that to some later point but if that is the decided course, then I would like to see the GitHub Issue number of that postponed work here before resolving the conv.